emails/tests/views_tests.py

```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
 
     @override_flag("developer_mode", active=True)
-    def test_complaint_mpp_3932(self):
-        """MPP-3932: Log notification for all complaints for developer_mode users."""
+    def test_complaint_developer_mode(self):
+        """Log complaint notification for developer_mode users."""
+
+        simulator_complaint_message = deepcopy(self.complaint_msg)
+        simulator_complaint_message["complaint"]["complainedRecipients"] = [
+            {"emailAddress": f"complaint+{self.ra.address}@simulator.amazonses.com"}
+        ]
+        complaint_body = {"Message": json.dumps(simulator_complaint_message)}
+
         with self.assertLogs(INFO_LOG) as logs:
-            response = _sns_notification(self.complaint_body)
+            response = _sns_notification(complaint_body)
         assert response.status_code == 200
 
         self.user.profile.refresh_from_db()
         assert self.user.profile.auto_block_spam is True
         self.mock_ses_client.send_raw_email.assert_not_called()
 
-        assert len(logs.records) == 2
-        record_mpp_3932 = logs.records[0]
-        assert record_mpp_3932.msg == "_handle_complaint: MPP-3932"
-        assert getattr(record_mpp_3932, "mask_id") == "unknown"
-        assert getattr(record_mpp_3932, "dev_action") == "log"
-        assert getattr(record_mpp_3932, "parts") == 1
-        assert getattr(record_mpp_3932, "part") == 0
-        notification_gza85 = getattr(record_mpp_3932, "notification_gza85")
+        (rec1, rec2) = logs.records
+        assert rec1.msg == "_handle_complaint: developer_mode"
+        assert getattr(rec1, "mask_id") == self.ra.metrics_id
+        assert getattr(rec1, "dev_action") == "log"
+        assert getattr(rec1, "parts") == 1
+        assert getattr(rec1, "part") == 0
+        notification_gza85 = getattr(rec1, "notification_gza85")
         log_complaint = decode_dict_gza85(notification_gza85)
-        expected_log_complaint = json.loads(self.complaint_body["Message"])
-        assert log_complaint == expected_log_complaint
+        assert log_complaint == simulator_complaint_message
 
-        record = logs.records[1]
-        assert record.msg == "complaint_notification"
+        assert rec2.msg == "complaint_notification"
+
+    @override_flag("developer_mode", active=True)
+    def test_complaint_developer_mode_missing_mask(self):
+        """If the simulator email can not be turned into a mask, it is not changed."""
+
+        simulator_complaint_message = deepcopy(self.complaint_msg)
+        simulator_complaint_message["complaint"]["complainedRecipients"] = [
+            {"emailAddress": "complaint+missing@simulator.amazonses.com"}
+        ]
+        complaint_body = {"Message": json.dumps(simulator_complaint_message)}
+
+        with (
+            self.assertLogs(INFO_LOG) as logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
```

Comment:
question (blocking): It looks like, unlike the code in `process_emails_from_sqs`, this complaint-handling code is run in the context of an HTTP callback request from AWS SNS. What does SNS do if it receives a 404 from Relay?

Reply:
The only documentation I can find is point 5 on https://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.prepare.html:

> 5. Make sure that your endpoint responds to the HTTP POST message from Amazon SNS with the appropriate status code. The connection will time out in approximately 15 seconds. If your endpoint does not respond before the connection times out, or if your endpoint returns a status code outside the range of 200–4xx, Amazon SNS will consider the delivery of the message as a failed attempt.

This suggests a 404 would be treated as a successful delivery. If logging is setup, the status code could be used to see how many complaints did not match current Relay users or masks.

Reply:
Thanks for digging that up! In this case I think `404` is fine.